### ISSUE TITLE
Questions on training

### ISSUE THREAD

**Issue by HuangDihe** [zju3dv/LoFTR#21]
Hi, thanks for your great work.
I tried to reproduce the training code, and then I met some problems:
1. I tried to set match threshold to 0.2 without loading pretrained model, and I found there will be 0 matches from coarse mathcing layer.
2. If I load the pretrained model to train, the Loss_coarse would go up while Loss_fine going down, and as training time goes on, the num of points from coarse matching layer reduces to 0.
3. Should I use gradient clipping?

**Comment by HuangDihe** [comment]
and what is EPI_ERR_THR in default.py ?

**Comment by angshine** [comment]
1 & 2. Training of the coarse-level LoFTR should converge pretty fast, so I guess there are bugs in the re-implementation, such as incorrect setup of gt coarse matches or loss calculation. Maybe Try overfitting one training sample before moving on.
3. We clip gradients with a norm above 0.5, but I think this should be irrelevant to the problem you met.
4. `EPI_ERR_THR` is the correctness threshold for a match w.r.t. the squared symmetric epipolar error.

**Comment by HuangDihe** [comment]
Thank you for your quick response!
I have visualize the gt coarse matches and they are all right.
I am confused about "pad coarse-level matches by randomly sampling from ground-truth matches", is this sampling on the coarse matching? or sample when calculating loss?

**Comment by HuangDihe** [comment]
> 1 & 2. Training of the coarse-level LoFTR should converge pretty fast, so I guess there are bugs in the re-implementation, such as incorrect setup of gt coarse matches or loss calculation. Maybe Try overfitting one training sample before moving on.
> 3. We clip gradients with a norm above 0.5, but I think this should be irrelevant to the problem you met.
> 4. `EPI_ERR_THR` is the correctness threshold for a match w.r.t. the squared symmetric epipolar error.

I pad coarse-level matches by randomly sampling from ground-truth matches after coarse mathcing, and as times go, there still is 0 match from coarse mathcing layer. All coarse matches are from gt matches. The coarse loss couldn't convergence. Could you  give me more details about training code?

**Comment by HuangDihe** [comment]
> 1 & 2. Training of the coarse-level LoFTR should converge pretty fast, so I guess there are bugs in the re-implementation, such as incorrect setup of gt coarse matches or loss calculation. Maybe Try overfitting one training sample before moving on.
> 3. We clip gradients with a norm above 0.5, but I think this should be irrelevant to the problem you met.
> 4. `EPI_ERR_THR` is the correctness threshold for a match w.r.t. the squared symmetric epipolar error.

I found that code 'mconf = conf_matrix[b_ids. i_ids, j_ids]', mconf don't have grad (mconf.requires_grad is False), conf_matrix.requires_grad is True.

**Comment by angshine** [comment]
The line `mconf = conf_matrix[b_ids, i_ids, j_ids]` extracts coarse-level matching confidences from the dense confidence matrix. We calculate the coarse loss based on the dense coarse-level matrix instead of the indexing results.

**Comment by HuangDihe** [comment]
> The line `mconf = conf_matrix[b_ids, i_ids, j_ids]` extracts coarse-level matching confidences from the dense confidence matrix. We calculate the coarse loss based on the dense coarse-level matrix instead of the indexing results.

Thank you very much!  I find the line '@torch.no_grad()' which causes the bugs. The coarse-level LoFTR should converge pretty fast, but the loss of fine level is very large.  Is this normal?  If I train loding pretrain model, both loss converge. 
And should I set the 'var.detach()'  ?

**Comment by HuangDihe** [comment]
> The line `mconf = conf_matrix[b_ids, i_ids, j_ids]` extracts coarse-level matching confidences from the dense confidence matrix. We calculate the coarse loss based on the dense coarse-level matrix instead of the indexing results.

Thank you very much! I find the line '@torch.no_grad()' which causes the bugs!
And one more question, should I set the 'var.detach()'  ?

**Comment by angshine** [comment]
Sorry, I don't quite get your question. If you are referring to whether to detach the indices for extraction of fine-level patches, then yes. However, in our case, it should already be decorated with `torch.no_grad()`.

**Comment by HuangDihe** [comment]
> Sorry, I don't quite get your question. If you are referring to whether to detach the indices for extraction of fine-level patches, then yes. However, in our case, it should already be decorated with `torch.no_grad()`.

Thank you.
